bash_completion

```diff
@@ -1209,108 +1209,102 @@ __parse_options()
     if [[ $option =~ (\[((no|dont)-?)\]). ]]; then
         option2=${option/"${BASH_REMATCH[1]}"/}
         option2=${option2%%[<{().[]*}
-        printf '%s\n' "${option2/=*/=}"
+        _options+=("${option2/=*/=}")
         option=${option/"${BASH_REMATCH[1]}"/"${BASH_REMATCH[2]}"}
     fi
 
     option=${option%%[<{().[]*}
     option=${option/=*/=}
     [[ $option ]] || return 1
 
-    printf '%s\n' "$option"
+    _options+=("$option")
 }
 
-# Parse GNU style help output of the given command.
-# @param $1  command; if "-", read from stdin and ignore rest of args
-# @param $2  command options (default: --help)
-#
-# TODO: rename per API conventions, rework to use vars rather than outputting
-_parse_help()
+# Parse GNU style help output of the given command and generate and store
+# completions in an array.
+# Usage:
+#   _comp_compgen_help [-u] command [opt]
+# Options:
+#   -u        Parse the BSD style usage output.  The default is in the GNU
+#             style help output.
+# @param $1 command  command; if "-", read from stdin and ignore rest of args
+# @param $2 opt      command options (default: --help)
```

Comment:
I'm wondering if passing everything after a `--` to the command would be a cleaner and more consistent API for this? For example `_comp_compgen` uses that approach.

Reply:
I also thought about it, but as far as the command name `$1` does not start with `-`, we can actually omit `--`. The difference in `_comp_compgen` is that the arguments to `compgen` always starts with `-`. But I think it is good to prefix `--` before every `"$1"`.

Reply:
Hmm, maybe it is better to split the function into two separate functions `_comp_compgen_help` and `_comp_compgen_usage`.

Before introducing the idea `_comp_compgen NAME`, I originally combined `_parse_help` and `_parse_usage` in a single function because it shared long codes of the command-line option parsing. But now the option parsing is moved to `_comp_compgen`, so two processings `help` and `usage` now share only a small part of the code. I'll think about it again.

Reply:
I split the function into `_comp_compgen_help` and `_comp_compgen_usage`.

Reply:
> Then again, I suppose we could make the first parameter optional too, and default it to `${comp_args[0]}`. If so, usage like `[-c command] [--] [command_opts...]` could be in order.

I like this idea. Let me implement it.

Reply:
After thinking about it again, I have slightly changed the suggested command line usage:

```bash
usage: _comp_compgen_help -
usage: _comp_compgen_help -c cmd args...
usage: _comp_compgen_help [[--] args...]
```

- When we want to read stdin, requiring to write `_comp_compgen_help -c -` seems strange, so `_comp_compgen_help -` is supported separately.
- When the caller specifies the command and arguments, splitting between `$1` and `$2...` with `--` as `_comp_compgen_help -c cmd -- args` also seems not so natural. So when the first argument is `-c`, the later words are all treated as the command and arguments as is.
- `_comp_dequote` is only applied to `comp_args[0]`. The command name specified through `-c cmd args...` is not de-quoted by `_comp_dequote`. When it is needed, the caller should take care of it.
- `--help` (or `--usage` in the case of `_comp_compgen_usage`) is only supplemented when no arguments are specified. E.g., when the user uses `_comp_compgen_help --`, it doesn't specify `--help` but directly calls `comp_args[0]` without arguments.

----

I rebased and modified existing commits. There are additional commits:

- There are additional fixes for `localvar_inherit`: 0a87c15e
- The backslash in `-\?` in `ssh-add`was removed: 01cfc44f
- These are non-trivial changes (where originally, a command name different from `"$1"` or multiple words in `$2` are supplied): fbe62ba5